### ISSUE TITLE
Why are you converting inline events to event listeners?

### ISSUE THREAD

**Issue by rbiggs** [hyperapp/hyperapp#163]
TL;DR
Treat inline events like attributes when diff/patching the DOM.

I've noticed you're having lots of issues with events, so I looked at the code. I noticed in the `setElementData` method you have the following:

```js
function setElementData(element, name, value, oldValue) {
  ...

  } else if (name[0] === "o" && name[1] === "n") {
    var event = name.substr(2)

    element.removeEventListener(event, oldValue)
    element.addEventListener(event, value)

  }
```
It's not clear to me why you would need to convert the inline events into event listeners. They are two different things. Inline events are treated like any other attributes on a element. You can set and remove them like so:

```js
// Attach inline event:
document.querySelector(‘#btn’).setAttribute(‘onclick’, ‘alert(“Hi!”)’)
// Clicking button will alert “Hi!”.
// Now remove inline event:
document.querySelector(‘#btn’).removeAttribute(‘onclick’)
// Now clicking button does nothing.
```
In fact, you can get the inline event on an element through its attributes property:

```html
<button onclick=‘alert(“Hello”)’>Click</button>
```
```js
document.querySelector(‘button’).attributes[0].firstChild === “onclick”
// true
```

I'm pretty sure what you're doing there may create duplicate events on documents, leading to various hard to sort out issues.

Or am I missing something here? (possibly)

**Comment by jbucaran** [comment]
@rbiggs Good point. I think it could even make the code simpler, specially since removing events (not in master) added some extra bytes.

**Comment by jbucaran** [comment]
@dodekeract @zaceno @lukejacksonn @terkelg What do you think?

**Comment by jbucaran** [comment]
> @rbiggs I'm pretty sure what you're doing there may create duplicate events on documents,

#160.

I fixed this locally by making sure the events are removed, but I haven't pushed yet, so I can still experiment with your suggestion before we create a new release.

**Comment by zaceno** [comment]
I agree that inline events should be treated like any other attribute, and can't think of any reasons why you'd convert them to addEventListener events

**Comment by jbucaran** [comment]
Great! I'm not 100% sure, but this refactoring will save us quite a few bytes!

**Comment by rbiggs** [comment]
Cool, check to see if it solves some of the other event-related issues 🙄

**Comment by jbucaran** [comment]
@rbiggs Yes, by using them as attributes, we would be overwriting the previous event, so I think cleaning up should be easier.

**Comment by rbiggs** [comment]
I brought this up in relation to #160 and #161. I'm expecting that this will resolve them.

**Comment by jbucaran** [comment]
Will be done for 0.8.0.

**Comment by terkelg** [comment]
Looks good to me. Reminds me of this little JS riddle: https://jsbin.com/yaloram/edit?html,js,output

**Comment by dodekeract** [comment]
@terkelg That's an interesting behavior. Makes sense once you get it though.

**Comment by leeoniya** [comment]
fyi, there are some events that cannot be attached inline. `transitionend` is one that i know off the top of my head. ask google for others. fwiw, domvm also does inline attach/detach to save on housekeeping/complexity. the special events can be bound via js from lifecycle hooks.

**Comment by rbiggs** [comment]
Those are the types of things you'd probably be handling in a subscription any way.

**Comment by dodekeract** [comment]
@leeoniya We should document this behavior properly then. I'd expect them to work.

**Comment by jbucaran** [comment]
@dodekeract You are absolutely right.

**Comment by jbucaran** [comment]
@rbiggs Thanks a lot! This saved us like 200 bytes, give or take. 🎉

**Comment by rbiggs** [comment]
And another bytes the dust 😉

**Comment by dodekeract** [comment]
@jbucaran We should at least consider this: https://en.wikipedia.org/wiki/Content_Security_Policy

Inline events may not be available in the environment. There are good reasons to use CSP. What do you think?

**Comment by jbucaran** [comment]
When are inline events not available?

**Comment by jbucaran** [comment]
I think you mean this? https://en.wikipedia.org/wiki/Content_Security_Policy#Mode_of_operation

> If the Content-Security-Policy header is present in the server response, a compliant client enforces the declarative whitelist policy. One example goal of a policy is a stricter execution mode for JavaScript in order to prevent certain cross-site scripting attacks. In practice this means that a number of features are disabled by default:
>> * DOM event handlers as HTML attributes (e.g. onclick)

**Comment by dodekeract** [comment]
Yes.

**Comment by jbucaran** [comment]
This means that if someone is serving a hyperapp and the server sets a Content-Security-Policy prop in header, we'd have to use lifecycle events to hook the events to the elements or events wouldn't work in our current implementation.

**Comment by jbucaran** [comment]
@dodekeract BTW, inline styles would also fail.

**Comment by jbucaran** [comment]
There seems to be some workarounds for some of the disabled stuff:
* https://en.wikipedia.org/wiki/Content_Security_Policy#cite_note-fn1-21

**Comment by rbiggs** [comment]
So, how does ELM and all the other libraries that use inline events handle that?

**Comment by jbucaran** [comment]
I'd like to know too, but I guess some may use addEventListener like were before. 

Working around Content-Security-Policy or going back to addEventListener is out of the question anyway.
